### PR TITLE
Reducing memory usage: removing useless logits computation in generate()

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

This is the PR related to the discussion in #30860. 
I followed was has been done in `Jamba` and added support for the `num_logits_to_keep` argument in `forward()`. However, even if this argument is `None`, the logits will only be upcasted to float if `labels` are passed (in order to accurately compute the loss). Otherwise, the upcasting only happen in the `generate()` functions.

For now, I only modified `Llama` and `Mistral`, but if you agree on the changes I will add support for more models.

## Benchmarks

Here I provide some benchmarks of the peak memory usage. For each input size, I generated 10 additional tokens.
Of course, since for few additional tokens the memory peak scales only with the first forward pass (at least when computing the whole logits matrix), and that the first forward scales linearly with input size and batch size (with new attention algorithms), the gain is actually constant for all input sizes and generation methods (except for contrastive search, which artificially increase the batch size after the first forward, thus the memory usage is slightly different). However, I still provide results for all generation methods here for completeness.

Basically we get: 
Llama3 8B -> MIND-BLOWING 3.62 memory usage reduction factor (due to large vocabulary)
Llama2 7B -> 1.17 reduction factor
Mistral 7B -> 1.32 reduction factor

Note that the memory reduction shown here is **on top** of whatever gains #30536 already provides for small new additional tokens, as I am comparing memory with the main transformers branch after it was merged. It integrates very nicely with that last PR, as the last one was providing most benefits when generating  more tokens, and this one provides gains for small new number of tokens.

[greedy.pdf](https://github.com/user-attachments/files/15645353/greedy.pdf)
[sample.pdf](https://github.com/user-attachments/files/15645374/sample.pdf)
[beam sample.pdf](https://github.com/user-attachments/files/15645320/beam.sample.pdf)
[beam search.pdf](https://github.com/user-attachments/files/15645334/beam.search.pdf)
[group beam search.pdf](https://github.com/user-attachments/files/15645364/group.beam.search.pdf)
[contrastive search.pdf](https://github.com/user-attachments/files/15645344/contrastive.search.pdf)

Here is a link to the benchmark script: https://gist.github.com/Cyrilvallez/92f48e402aa2968c854a8128796f50c3

## Who can review?

@ArthurZucker @gante Let me know what you think about the proposed changes!

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @ArthurZucker @gante

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1446,6 +1446,13 @@ def _supports_default_dynamic_cache(self) -> bool:
         """
         return self._supports_cache_class and "jamba" not in self.__class__.__name__.lower()
 
+    def _supports_num_logits_to_keep(self) -> bool:
+        """
+        Return True if the current model supports the keyword argument `num_logits_to_keep` in forward()
+        to save memory. Checking it in this way allows to avoid using a new model attribute.
+        """
+        return "num_logits_to_keep" in set(inspect.signature(self.forward).parameters.keys())
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1454]: Mmmmm I don't think we will add it to the signature, let's rather wait for this PR to be merged #31446 which will add kwargs

**Review by ArthurZucker — COMMENTED** [review]
Thanks! Looks quite alright IMO!

File: src/transformers/models/mistral/modeling_mistral.py [hunk 2]
```diff
@@ -1202,11 +1208,16 @@ def forward(
         )
 
         hidden_states = outputs[0]
-        logits = self.lm_head(hidden_states)
-        logits = logits.float()
+        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        if num_logits_to_keep is None:
+            logits = self.lm_head(hidden_states)
+        else:
+            logits = self.lm_head(hidden_states[:, -num_logits_to_keep:, :])
```
Comment (gante) [inline src/transformers/models/mistral/modeling_mistral.py:1215]: Note on BC: despite [torch's type promotion](https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc), this change can be breaking on external scripts. For instance, if a user decides to add a FP16 tensor to `logits`, previously it would accumulate in FP32 and now it can accumulate in FP16.

Because of this, I'd suggest going through a deprecation cycle, e.g.
```py
        if labels is None:
            logger.warning_once(
                'Starting from v4.44, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)'
            )
        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
        if num_logits_to_keep is None:
            logits = self.lm_head(hidden_states).float()
        else:
            logits = self.lm_head(hidden_states[:, -num_logits_to_keep:, :]).float()
```

Then on v4.44 we would remove the `.float()`, giving the users time to adapt :)

**Review by gante — COMMENTED** [review]
LGTM, thank you for further reducing the memory needs @Cyrilvallez 💛 

`num_logits_to_keep` is not the prettiest interface, but I can't think of a better one (as discussed in the PR that introduced it). 

I'm happy with the PR with the exception of BC handling

**Comment by gante** [comment]
btw, a ratio of 3x lower peak memory consumption is 🔥 🔥 🔥

File: src/transformers/generation/utils.py (hunk 1 above)
Comment (Cyrilvallez) [reply src/transformers/generation/utils.py:1454]: I see the related PR only adding kwargs to the forward of the attention blocks, not to the forward of the main ModelForCausalLM. Will it be added in the future or am I missing something?

**Review by gante — APPROVED** [review]
LGTM 👍

**Comment by Cyrilvallez** [comment]
I just added the change to more models and rebased to avoid conflicts with new commits in main!
For Cohere-based models, I most notably computed a **memory gain ratio of 6.68** due to the very large 256k vocabulary size 🚀🔥

File: src/transformers/generation/utils.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/generation/utils.py:1458]: Ignore what I said, as this is high impact feature, we can add it to the signature

File: src/transformers/models/cohere/modeling_cohere.py [hunk 3]
```diff
@@ -1126,12 +1132,22 @@ def forward(
...
+        if labels is None:
+            logger.warning_once(
+                "Starting from v4.44, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)"
+            )
+        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        # TODO: remove those 2 float() operations in v4.44
+        if num_logits_to_keep is None:
+            logits = self.lm_head(hidden_states).float()
+        else:
+            logits = self.lm_head(hidden_states[:, -num_logits_to_keep:, :]).float()
```
Comment (ArthurZucker) [inline src/transformers/models/cohere/modeling_cohere.py:1144]: we can default num_logits to keep to 0 to always slice (no codepaths)

**Review by ArthurZucker — COMMENTED** [review]
Overall LGTM

**Comment by ArthurZucker** [comment]
Make sure to rebase as the state of the main branch was changed quite a bit!

**Comment by Cyrilvallez** [comment]
DO NOT MERGE YET
Everything else is good, but still need to sort out the logger.warning_once/compile issue

**Comment by Cyrilvallez** [comment]
@ArthurZucker @gante everything is now ready.
From my tests, it seems like `compile` does not support any print-like functionality at the moment, either from `print`, `logger` or `warnings`.
I first wanted to add a `logger.warning_once_compile_safe` function which I thought would simplify things and come in handy in the future as well, but couldn't because it needs to import `torch` in the `logging` module which breaks things. 
So I just added a compile check everywhere.

**Comment by ringohoffman** [comment]
@ArthurZucker is this planned for review this week? I’m pretty eager to consume this PR.

**Comment by ArthurZucker** [comment]
Yes!

**Comment by Oxi84** [comment]
Looking forward to testing this out, gemma2 uses a lot of memory  otherwise and is a top model.

File: src/transformers/models/cohere/modeling_cohere.py [hunk 4]
```diff
@@ -1021,12 +1028,19 @@ def forward(
         )
 
         hidden_states = outputs[0]
-        logits = self.lm_head(hidden_states)
+        if labels is None and not is_torchdynamo_compiling():
+            logger.warning_once(
+                "Starting from v4.44, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)"
```
Comment (ArthurZucker) [inline src/transformers/models/cohere/modeling_cohere.py:1033]:
```suggestion
                "Starting from v4.45, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)"
```

File: src/transformers/generation/utils.py
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1]: Maybe let's comment that we need .float() for full precision soft

**Review by ArthurZucker — APPROVED** [review]
Well LGTM one thing missing is a test in the mixins !

**Comment by ringohoffman** [comment]
Hey @Cyrilvallez, thanks for your work. Just checking in regarding this PR. Do you have a plan to finish it up some time soon?  I'm very excited for it to land!

**Comment by Cyrilvallez** [comment]
Hi @ringohoffman, don't worry I am not forgetting about this 😉 I'm currently on vacation so I will try to wrap it up quickly end of August when I come back if I have time. Worst case scenario, it will be ready mid-September.

In the meantime, you can install transformers from my fork if you want to already benefit from it (pip install git+https://github.com/Cyrilvallez/transformers@logits-dtype). Or even better, you can clone my fork and rebase it on transformers/main to get all the new stuff + this PR.

**Comment by Boubou78000** [comment]
Does this PR actually fixes gemma2 or just Gemma?

**Comment by Cyrilvallez** [comment]
Gemma2 was not released yet when I started this, but don't worry I will add it as well, it's on the roadmap 🤗

**Comment by Cyrilvallez** [comment]
@ArthurZucker I added support for Gemma2 as well as tests, ready for last check 🤗
Red CIs are not related to the PR

**Comment by ArthurZucker** [comment]
Can you rebase to fix the CI? 🤗

**Comment by Cyrilvallez** [comment]
Fixed it!
Last red CI is the `libtorch_cuda.so: cannot open shared object file: No such file or directory` issue that was also present on recent PRs.

**Comment by gante** [comment]
@Cyrilvallez the CI images built after [this PR](https://github.com/huggingface/transformers/pull/32918) gets merged should fix the CI issue we're seeing 🤗

**Comment by Cyrilvallez** [comment]
Indeed it worked, @ArthurZucker everything is now green!

File: src/transformers/models/cohere/modeling_cohere.py [hunk 5]
```diff
@@ -1071,12 +1078,19 @@ def forward(
         )
 
         hidden_states = outputs[0]
-        logits = self.lm_head(hidden_states)
+        if labels is None and not is_torchdynamo_compiling():
+            logger.warning_once(
+                "Starting from v4.45, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)"
```
Comment (ArthurZucker) [inline src/transformers/models/cohere/modeling_cohere.py:1083]:
```suggestion
                "Starting from v4.46, the `logits` model output will have the same type as the model (except at train time, where it will always be FP32)"
```

**Review by ArthurZucker — APPROVED** [review]
LGTM!

File: src/transformers/models/cohere/modeling_cohere.py (hunk 5 above)
Comment (ArthurZucker) [reply src/transformers/models/cohere/modeling_cohere.py:1083]: should be changed everywhere as next version is 4.45!

Comment (Cyrilvallez) [reply src/transformers/models/cohere/modeling_cohere.py]: Done but the `examples_tensorflow` test started to fail again on the CI 🥲

**Comment by gante** [comment]
@Cyrilvallez rebasing (again) will get rid of the CI error

(bear with us, we're chasing these CI blockers 🙏 )

**Comment by Cyrilvallez** [comment]
No worries! All good on the CIs and ready to be merged 🤗
